message_flag_updates: Fix handling of post failures for flag updates.

roanster007 opened the pull request:
Previously, when a message's POST of flag update fails, the UI is rendered as if it has succeeded. However, when the site is reloaded, the change could not be seen (for example starring of messages).

This is fixed by storing the POST failures in localstorage,and updating the it when the site reloads and flushing the values which were updated from the localstorage.

<!-- Describe your pull request here.-->

[CZO Thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Message.20not.20getting.20added.20to.20the.20Starred.20messages.2E)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Testing Procedure** 
- Terminate the Dev Server.
- Star some message.
- Restart the server.
- The Starred message could still be seen in the Starred list.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/message_flags.js:
```diff
@@ -1,17 +1,32 @@
+                    ls.set("untrackedFlaggedMessages", [{msg_ids, flag, op}]);
+                }
+            }
+        });
```
Zulip does not use `camelCase`, please follow our coding conventions.

timabbott commented on web/src/ui_init.js:
```diff
@@ -435,6 +436,15 @@ export function initialize_everything() {
+            message_flags.send_flag_update_for_messages(message.msg_ids, message.flag, message.op);
+        }
+        ls.remove("untrackedFlaggedMessages");
+    }
```
Please don't put new code inline in `ui_init.js`; that module should call a function in the `message_flags` module that contains this logic, to have reasonable abstractions.

Also, we likely certainly shouldn't do this as the very first thing before other initialization is completed.

timabbott commented:
Please take more time reading our style guide and also just reading the surrounding code and coding conventions when making changes; it's unpleasant to review pull requests that don't follow our guidelines. For contributing to a large OSS project like Zulip, it's far far more important how nice you make it to review your work than how many PRs you submit.

The other big concern I have with this is that it likely needs an expiry period of no more than a week; I don't think one would want stale values in local storage from 6 months ago on an old laptop to be sent when booting said computer.

(Also, does this correctly handle the case that multiple requests have failed?)

timabbott commented on web/src/message_flags.js:
```diff
@@ -1,17 +1,32 @@
+                op,
+            },
+        })
+        ?.fail(() => {
```
This is not how you check for failures in the `channel` module; you pass an `error` hook. The fact that you wrote this code this way suggests to me that you didn't test this. You should be testing, say by stopping your server so that all requests fail.

roanster007 replied:
@timabbott , I tested it with ```fail``` section, and it indeed worked. The reference for this section : [docs](https://api.jquery.com/deferred.fail/).  I also found out that the ```error``` hook was deprecated with ```fail``` for ajax requests.

timabbott replied:
Ahh, OK. The `channel` module isn't designed to support that with its error wrapper logic; though maybe we should change it, but I think if we care we might make `channel` use `fail` internally.

roanster007 replied:
Ya, actually that time when I wrote the code, I didn't have the idea of how it was done.... I'll update it.

roanster007 commented:
I am really sorry regarding the inconvenience. I will update the PR as soon as possible.

roanster007 commented:
@timabbott , I have updated the PR as per your review. I have also uploaded the screencast of working.

- In addition, I have also updated in considering the situation where the POST fails during reload to avoid duplication of the same thing while reloading.
- Regarding the Branch coverage test, I couldn't find any reference or a way to mock the failure of POST. So I have not made any test for it. (should it be exempted for code coverage ?)
- Regarding the use of ```fail()``` rather than ```error``` hook, I have answered it in the review.

timabbott commented:
`web/tests/example7.test.js` has an example of mocking HTTP post requests that may be helpful?

timabbott commented:
Closing in favor of #35559 which completes this work.

timabbott commented:
Actually I'm not entirely sure I buy the strategy for #35559, reopening pending having more time to review the options.
